components/webdriver_server/actions.rs

```diff
@@ -24,11 +25,91 @@ use crate::{Handler, WebElement, wait_for_script_response};
 static POINTERMOVE_INTERVAL: u64 = 17;
 static WHEELSCROLL_INTERVAL: u64 = 17;
 
+// A single action, corresponding to an `action object` in the spec.
+pub(crate) enum ActionItem {
+    Null(NullActionItem),
+    Key(KeyActionItem),
+    Pointer(PointerActionItem),
+    Wheel(WheelActionItem),
+}
+
+// A set of actions with multiple sources executed within a single tick.
+// The order in which they are performed is not guaranteed.
+// The `id` is used to identify the source of the actions.
+pub(crate) struct TickActions {
+    pub actions: HashMap<String, ActionItem>,
+}
+
+// Consumed by the `dispatch_actions` method.
+pub(crate) struct ActionsByTick {
+    pub actions: Vec<TickActions>,
+}
+
+/// <https://w3c.github.io/webdriver/#dfn-extract-an-action-sequence>
+pub(crate) fn extract_an_action_sequence(params: ActionsParameters) -> ActionsByTick {
+    // Step 1. Let actions be the result of getting a property named "actions" from parameters.
+    let actions = params.actions;
+
+    actions_by_tick_from_sequence(actions)
+}
+
+pub(crate) fn actions_by_tick_from_sequence(actions: Vec<ActionSequence>) -> ActionsByTick {
+    let mut actions_by_tick: ActionsByTick = ActionsByTick {
+        actions: Vec::new(),
+    };
+
+    // Step 4. For each value action sequence corresponding to an indexed property in actions
+    for action_sequence in actions {
+        let mut source_actions: Vec<ActionItem> = Vec::new();
+
+        match action_sequence.actions {
+            ActionsType::Null {
+                actions: null_actions,
+            } => {
+                source_actions.extend(null_actions.into_iter().map(ActionItem::Null));
+            },
+            ActionsType::Key {
+                actions: key_actions,
+            } => {
+                source_actions.extend(key_actions.into_iter().map(ActionItem::Key));
+            },
+            ActionsType::Pointer {
+                parameters: _,
+                actions: pointer_actions,
+            } => {
+                source_actions.extend(pointer_actions.into_iter().map(ActionItem::Pointer));
+            },
+            ActionsType::Wheel {
+                actions: wheel_actions,
+            } => {
+                source_actions.extend(wheel_actions.into_iter().map(ActionItem::Wheel));
+            },
+        }
```

Comment:
Could you move this to a dedicated function `process_an_input_source_action_sequence` to better match the spec?